Feature: Dominant Eigenvalue Estimator (#750)
LLNL/sundials: doc/shared/sundomeigest/SUNDomEigEst_API.rst

   SUNDIALS Copyright Start
   Copyright (c) 2002-2025, Lawrence Livermore National Security
   and Southern Methodist University.
   All rights reserved.

   See the top-level LICENSE and NOTICE files for details.

   SPDX-License-Identifier: BSD-3-Clause
   SUNDIALS Copyright End
   ----------------------------------------------------------------

.. _SUNDomEigEst.API:

The SUNDomEigEstimator API
=============================

The SUNDomEigEst API defines several dominant eigenvalue estimation operations that enable
SUNDIALS packages to utilize this API.  These functions can be divided into three categories.
The first are the core dominant eigenvalue estimation functions.  The second consist of "set"
routines to supply the dominant eigenvalue estimator with functions provided by the SUNDIALS
packages and to modify estimator parameters.  The final group consists of "get" routines for
retrieving dominant eigenvalue estimation statistics.  All of these functions are defined in
the header file ``sundials/sundials_domeigestimator.h``.

.. _SUNDomEigEst.CoreFn:

SUNDomEigEstimator core functions
-----------------------------------------------------

> The core dominant eigenvalue estimator functions consist of several **required**
> functions: :c:func:`SUNDomEigEst_NewEmpty` creates a new empty ``SUNDomEigEstimator`` 
> object, :c:func:`SUNDomEigEst_SetATimes` provides a :c:type:`SUNATimesFn` function pointer,

Reviewer: I wouldn't label `SUNDomEigEst_NewEmpty` as a "required" function, since it is a utility routine provided to implementors (who never need to write this themselves).  I recommend something more like:
```suggestion
The SUNDomEigEstimator base class provides two **utility** routines for implementors, :c:func:`SUNDomEigEst_NewEmpty` and :c:func:`SUNDomEigEst_FreeEmpty`.

Implementations of SUNDomEigEstimators must include a set of **required** functions: `
:c:func:`SUNDomEigEst_SetATimes` provides a :c:type:`SUNATimesFn` function pointer,
```

as well as a ``void*`` pointer to a data structure used by this routine,

> :c:func:`SUNDomEigEst_Initialize` initializes the estimator object once all estimator-specific
> options have been set, :c:func:`SUNDomEig_Estimate` estimates the dominant eigenvalue,
> :c:func:`SUNDomEigEst_FreeEmpty` an empty estimator and :c:func:`SUNDomEigEst_Destroy` destroys 
> an estimator object.

Reviewer: ```suggestion
:c:func:`SUNDomEigEst_Initialize` initializes the estimator object once all estimator-specific
options have been set, :c:func:`SUNDomEig_Estimate` estimates the dominant eigenvalue,
and :c:func:`SUNDomEigEst_Destroy` destroys an estimator object.
```

.. c:function:: SUNDomEigEstimator SUNDomEigEst_NewEmpty(SUNContext sunctx)

>    *Required.*
> 

Reviewer: ```suggestion
```

   This function allocates a new ``SUNDomEigEstimator`` object and
   initializes its content pointer and the function pointers in the operations
   structure to ``NULL``.

   **Arguments:**

      * *sunctx* -- the :c:type:`SUNContext` object (see :numref:`SUNDIALS.SUNContext`)

   **Return value:**

      If successful, this function returns a ``SUNDomEigEstimator`` object.
      If an error occurs when allocating the object, then this routine will
      return ``NULL``.

.. c:function:: SUNErrCode SUNDomEigEst_SetATimes(SUNDomEigEstimator DEE, void* A_data, SUNATimesFn ATimes)

   *Required.*

   Provides a :c:type:`SUNATimesFn` function pointer, as well as a ``void*`` pointer to a
   data structure used by this routine, to the dominant eigenvalue estimator object
   *DEE*.  SUNDIALS packages call this function to set the matrix-vector product function
   to either an estimator-provided difference-quotient via vector operations or a user-supplied
   estimator-specific routine.

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object,

Reviewer: Since the object name is "SUNDomEigEstimator", but the function prefix is "SUNDomEigEst", then I recommend that these function descriptions use the correct object name instead of the shortened function prefix: 
```suggestion
      * *DEE* -- a SUNDomEigEstimator object,
```

      * *A_data* -- pointer to structure for ``ATimes``,
      * *ATimes* -- function pointer to perform :math:`Av` product.

   **Return value:**

      A :c:type:`SUNErrCode`.

   **Usage:**

      .. code-block:: c

         retval = SUNDomEigEst_SetATimes(DEE, A_data, ATimes);

.. c:function:: SUNErrCode SUNDomEigEst_Initialize(SUNDomEigEstimator DEE)

   *Required.*

   Performs dominant eigenvalue estimator initialization (assuming that all
   estimator-specific options have been set).

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object.

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object.
```

   **Return value:**

      A :c:type:`SUNErrCode`.

   **Usage:**

      .. code-block:: c

         retval = SUNDomEigEst_Initialize(DEE);

.. c:function:: SUNErrCode SUNDomEig_Estimate(SUNDomEigEstimator DEE, sunrealtype* lambdaR, sunrealtype* lambdaI)

   This *required* function estimates the dominant eigenvalue,
   :math:`\lambda_{\max} = \lambda_{R} + \lambda_{I}i` such that
   :math:`|\lambda| = \max\{|\lambda_i| : A \vec{v_i} = \lambda_i \vec{v_i}, \ \vec{v_i} \neq \vec{0} \}`.

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object.

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object.
```

      * *lambdaR* -- The real part of the dominant eigenvalue
      * *lambdaI* -- The imaginary part of the dominant eigenvalue

   **Return value:**

      `SUN_SUCCESS` for a successful call, or a relevant error code from
      :numref:`SUNDomEigEst.ErrorCodes` upon failure.

   **Usage:**

      .. code-block:: c

         retval = SUNDomEig_Estimate(DEE, dom_eig);

.. c:function:: SUNErrCode SUNDomEigEst_FreeEmpty(SUNDomEigEstimator DEE)

   This routine frees the ``SUNDomEigEstimator`` object, under the
   assumption that any implementation-specific data that was allocated
   within the underlying content structure has already been freed.
   It will additionally test whether the ops pointer is ``NULL``,
   and, if it is not, it will free it as well.

   **Arguments:**

      * *DEE* -- a SUNDomEigEstimator object

   **Return value:**

      A :c:type:`SUNErrCode`.

.. c:function:: SUNErrCode SUNDomEigEst_Destroy(SUNDomEigEstimator* DEEptr)

>    Frees memory allocated by the dominant eigenvalue estimatimator.

Reviewer: ```suggestion
   Frees memory allocated by the dominant eigenvalue estimator.
```

   **Arguments:**

>       * *DEEptr* -- a SUNDomEigEst object pointer.

Reviewer: ```suggestion
      * *DEEptr* -- a SUNDomEigEstimator object pointer.
```

   **Usage:**

      .. code-block:: c

         retval = SUNDomEigEst_Destroy(&DEE);

.. _SUNDomEigEst.SetFn:

SUNDomEigEstimator "set" functions
-------------------------------------

The following functions supply dominant eigenvalue estimator modules with
functions defined by the SUNDIALS packages and modify estimator parameters.
Only the routine for setting the matrix-vector product routine is required.
Otherwise, all other set functions are optional.  SUNDomEigEst implementations
that do not provide the functionality for any optional routine should leave the corresponding
function pointer ``NULL`` instead of supplying a dummy routine.

.. c:function:: SUNErrCode SUNDomEigEst_SetNumPreProcess(SUNDomEigEstimator DEE, int numpreprocess)

   This *optional* routine should set the number of "warm-up" matrix-vector multiplications,
   which is executed by :c:func:`SUNDomEig_Estimate` before each estimate.

> 

Reviewer: I'm confused.  I thought that `LSRKStepSetNumSucceedingWarmups` would set the number of warmups to be used before each call to `SUNDomEig_Estimate`, so I anticipated that `SUNDomEigEst_SetNumPreProcess` would set the number of warmup iterations performed only before the very **first** call to `SUNDomEig_Estimate`.  However, this documentation seems to imply that *both* functions warm up every time, and neither is done only at the very beginning.

Is the documentation for one of these functions incorrect, or do both perform warmup iterations before each `SUNDomEig_Estimate`?  If both do this, then one should be removed, or converted to a one-time warmup phase at the start of the simulation.

Author: Similar to my comments on the other spots, this is a true statement as a stand alone module. However, integrator changes this number after the very first estimate call (set to default 0 unless the user calls relevant LSRK set routine). We need to update this part and maybe add the same explanation that we will have for the other spots.

Author: I had to revert it since it was failing in the single setup and I couldn't figure out what was wrong. I think we can delay these changes to allow more time for thorough investigation.

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object,

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object,
```

      * *numpreprocess* -- the number of preprocess.

   **Return value:**

      * *numpreprocess* -- the number of preprocessing iterations. 

   **Usage:**

      .. code-block:: c

         retval = SUNDomEigEst_SetNumPreProcess(DEE, numpreprocess);

> 

Author: Should I remove the following note?

.. c:function:: SUNErrCode SUNDomEigEst_SetTol(SUNDomEigEstimator DEE, sunrealtype tol)

>    This *optional* routine sets the estimator's :ref:`relative tolerance <pi_rel_tol>`.

Reviewer: Here, and with many of the following routines:

Is this function callable by both "Power" and "Arnoldi," or just "Power"?  If only "Power" supports it, then I recommend removing it from the base class and putting it only inside "Power".

Author: Based on our discussion in the group meeting today, I will resolve this comment.

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object,

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object,
```

      * *tol* -- the requested eigenvalue accuracy.

   **Return value:**

      * *tol* -- the requested eigenvalue accuracy.

   **Usage:**

      .. code-block:: c

         retval = SUNDomEigEst_SetTol(DEE, tol);

.. c:function:: SUNErrCode SUNDomEigEst_SetMaxIters(SUNDomEigEstimator DEE, long int max_iters)

   This *optional* routine sets the maximum number of iterations.

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object,

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object,
```

      * *max_iters* -- the maximum number of iterations.

   **Return value:**

      A :c:type:`SUNErrCode`.

   **Usage:**

      .. code-block:: c

         retval = SUNDomEigEst_SetMaxIters(DEE, max_iters);

.. _SUNDomEigEst.GetFn:

SUNDomEigEstimator "get" functions
----------------------------------

The following functions allow SUNDIALS packages to retrieve results from a
dominant eigenvalue estimator.  *All routines are optional.*

.. c:function:: SUNErrCode SUNDomEigEst_GetCurRes(SUNDomEigEstimator DEE, sunrealtype* cur_res)

   This *optional* routine should return the final residual from
   the most-recent "estimator" call.

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object.

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object.
```

      * *cur_res* -- the current residual

   **Return value:**

      A :c:type:`SUNErrCode`.

   **Usage:**

      .. code-block:: c

         sunrealtype cur_res;
         retval = SUNDomEigEst_GetCurRes(DEE, &cur_res);

.. c:function:: SUNErrCode SUNDomEigEst_GetCurNumIters(SUNDomEigEstimator DEE, long int* curniter)

   This *optional* routine should return the number of estimator
   iterations performed in the most-recent "estimator" call.

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object,

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object,
```

      * *curniter* -- the current number of iterations.

   **Return value:**

      A :c:type:`SUNErrCode`.

   **Usage:**

      .. code-block:: c

         long int curniter;
         retval = SUNDomEigEst_GetCurNumIters(DEE, &curniter);

> .. c:function:: SUNErrCode SUNDomEigEst_GetMaxNumIters(SUNDomEigEstimator DEE, long int* max_niter)
> 
>    This *optional* routine should return the maximum number of iterations
>    performed in all "estimator" calls so far.

Reviewer: Is this the maximum number of iterations in any single "Estimate" call so far, or the total number of iterations used in all "Estimate" calls so far?  If its the former, then I recommend:
```suggestion
.. c:function:: SUNErrCode SUNDomEigEst_GetMaxNumIters(SUNDomEigEstimator DEE, long int* max_niter)

   This *optional* routine should return the maximum number of iterations
   performed in any previous "estimator" call so far.
```

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object,

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object,
```

      * *max_niter* -- the maximum number of iterations.

   **Return value:**

      A :c:type:`SUNErrCode`.

   **Usage:**

      .. code-block:: c

         long int max_niter;
         retval = SUNDomEigEst_GetMaxNumIters(DEE, &max_niter);

.. c:function:: SUNErrCode SUNDomEigEst_GetMinNumIters(SUNDomEigEstimator DEE, long int* min_niter)

   This *optional* routine should return the minimum number of iterations

>    performed in all "estimator" calls so far.

Reviewer: ```suggestion
   performed in any previous "estimator" call so far.
```

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object,

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object,
```

      * *min_niter* -- the minimum number of iterations.

   **Return value:**

      A :c:type:`SUNErrCode`.

   **Usage:**

      .. code-block:: c

         long int min_niter;
         retval = SUNDomEigEst_GetMinNumIters(DEE, &min_niter);

.. c:function:: SUNErrCode SUNDomEigEst_GetNumATimesCalls(SUNDomEigEstimator DEE, long int* num_ATimes)

   This *optional* routine should return the number of calls to the :c:type:`SUNATimesFn` function.

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object,

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object,
```

      * *num_ATimes* -- the number of calls to the ``Atimes`` function.

   **Return value:**

      A :c:type:`SUNErrCode`.

   **Usage:**

      .. code-block:: c

         long int num_ATimes;
         retval = SUNDomEigEst_GetNumATimesCalls(DEE, &num_ATimes);

.. c:function:: SUNErrCode SUNDomEigEst_PrintStats(SUNDomEigEstimator DEE, FILE* outfile)

   This *optional* routine prints the dominant eigenvalue estimator statistics
   to the output stream *outfile*.

   **Arguments:**

>       * *DEE* -- a SUNDomEigEst object,

Reviewer: ```suggestion
      * *DEE* -- a SUNDomEigEstimator object,
```

      The function implementing :c:func:`SUNDomEigEst_SetTol`

   .. c:member:: SUNErrCode (*initialize)(SUNDomEigEstimator)

      The function implementing :c:func:`SUNDomEigEst_Initialize`

   .. c:member:: SUNErrCode (*estimate)(SUNDomEigEstimator, sunrealtype*, sunrealtype*)

      The function implementing :c:func:`SUNDomEig_Estimate`

   .. c:member:: sunrealtype (*getcurres)(SUNDomEigEstimator)

      The function implementing :c:func:`SUNDomEigEst_GetCurRes`

   .. c:member:: int (*getcurniters)(SUNDomEigEstimator)

      The function implementing :c:func:`SUNDomEigEst_GetCurNumIters`

   .. c:member:: int (*getmaxniters)(SUNDomEigEstimator)

      The function implementing :c:func:`SUNDomEigEst_GetMaxNumIters`

   .. c:member:: int (*getminniters)(SUNDomEigEstimator)

      The function implementing :c:func:`SUNDomEigEst_GetMinNumIters`

   .. c:member:: long int (*getnumatimescalls)(SUNDomEigEstimator)

      The function implementing :c:func:`SUNDomEigEst_GetNumATimesCalls`
      
   .. c:member:: SUNErrCode (*printstats)(SUNDomEigEstimator, FILE*)

      The function implementing :c:func:`SUNDomEigEst_PrintStats`
      

>    .. c:member:: SUNErrCode (*free)(SUNDomEigEstimator)

Reviewer: Typically, the function pointer name matches the public API name, and this forgot a `*`
```suggestion
   .. c:member:: SUNErrCode (*destroy)(SUNDomEigEstimator*)
```

   +------------------------------------+-------+---------------------------------------------------+
   | ``SUN_ERR_DEE_NULL_CONTENT``       | -9967 | the DEE content is ``NULL``                       |
   +------------------------------------+-------+---------------------------------------------------+
   | ``SUN_ERR_DEE_LAPACK_FAIL``        | -9966 | LAPACK ``_dgeev/_sgeev`` function failure         |
   |                                    |       |                                                   |
   +------------------------------------+-------+---------------------------------------------------+
   | ``SUN_ERR_DEE_NULL_ESTIMATE``      | -9965 | estimate function ptr is ``NULL``                 |
   |                                    |       |                                                   |
   +------------------------------------+-------+---------------------------------------------------+
   | ``SUN_ERR_DEE_NULL_FREE``          | -9964 | free function ptr is ``NULL``                     |
   |                                    |       |                                                   |

>    +------------------------------------+-------+---------------------------------------------------+

Reviewer 2: `SUN_ERR_DEE_ATIMES_FAIL_REC` and `SUN_ERR_DEE_ATIMES_FAIL_UNREC` can be replaced with `SUN_ERR_USER_FCN_FAIL` (suggested this elsewhere) since recovery from a failed `Estimate` call is not possible.

>    +------------------------------------+-------+---------------------------------------------------+

Author: I have removed the table since there was nothing left after the updated errorcode returns.
